// @convostack/shared/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  AI_ROLE: () => AI_ROLE,
  CONVERSATION_EVENT_KIND_CONVERSATION_METADATA: () => CONVERSATION_EVENT_KIND_CONVERSATION_METADATA,
  CONVERSATION_EVENT_KIND_MESSAGE: () => CONVERSATION_EVENT_KIND_MESSAGE,
  CONVERSATION_EVENT_KIND_MESSAGE_PART: () => CONVERSATION_EVENT_KIND_MESSAGE_PART,
  HUMAN_ROLE: () => HUMAN_ROLE,
  addSecondsToDate: () => addSecondsToDate,
  generateRandomID: () => generateRandomID,
  unixTimestamp: () => unixTimestamp
});
module.exports = __toCommonJS(src_exports);

// src/consts.ts
var CONVERSATION_EVENT_KIND_MESSAGE = "message";
var CONVERSATION_EVENT_KIND_MESSAGE_PART = "message_part";
var CONVERSATION_EVENT_KIND_CONVERSATION_METADATA = "conversation_metadata";
var AI_ROLE = "AI";
var HUMAN_ROLE = "Human";

// src/random.ts
var import_crypto = __toESM(require("crypto"));
var generateRandomID = (length) => {
  return import_crypto.default.randomBytes(length).toString("hex");
};

// src/timestamps.ts
var unixTimestamp = (date) => {
  if (!date) {
    date = /* @__PURE__ */ new Date();
  }
  return Math.floor(date.getTime() / 1e3);
};
var addSecondsToDate = (date, seconds) => {
  const newUnixTimestamp = unixTimestamp(date) + seconds;
  return new Date(newUnixTimestamp * 1e3);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AI_ROLE,
  CONVERSATION_EVENT_KIND_CONVERSATION_METADATA,
  CONVERSATION_EVENT_KIND_MESSAGE,
  CONVERSATION_EVENT_KIND_MESSAGE_PART,
  HUMAN_ROLE,
  addSecondsToDate,
  generateRandomID,
  unixTimestamp
});
