// @convostack/frontend-react/dist/ConvoStackEmbed-a887c306.js
import { j as e, u as A, a as D, b as q, c as R, d as K, e as U, s as V, g as P, h as _, i as B } from "./index-90257461.js";
import { useContext as W, useState as j, useRef as M, useEffect as m } from "react";
import { C as z, T as G, S as Q, P as J, L as $, a as O, b as X, c as Y } from "./App-0e376667.js";
import Z from "./Message-9e7c431a.js";
import "react-dom";
const y = ({ className: a = "" }) => /* @__PURE__ */ e.jsx("div", { className: a, children: /* @__PURE__ */ e.jsx("div", { className: "border border-blue-300 shadow rounded-md p-4 max-w-sm w-full mx-auto", children: /* @__PURE__ */ e.jsxs("div", { className: "animate-pulse flex space-x-4", children: [
  /* @__PURE__ */ e.jsx("div", { className: "rounded-full bg-slate-700 h-10 w-10" }),
  /* @__PURE__ */ e.jsxs("div", { className: "flex-1 space-y-6 py-1", children: [
    /* @__PURE__ */ e.jsx("div", { className: "h-2 bg-slate-700 rounded" }),
    /* @__PURE__ */ e.jsxs("div", { className: "space-y-3", children: [
      /* @__PURE__ */ e.jsxs("div", { className: "grid grid-cols-3 gap-4", children: [
        /* @__PURE__ */ e.jsx("div", { className: "h-2 bg-slate-700 rounded col-span-2" }),
        /* @__PURE__ */ e.jsx("div", { className: "h-2 bg-slate-700 rounded col-span-1" })
      ] }),
      /* @__PURE__ */ e.jsx("div", { className: "h-2 bg-slate-700 rounded" })
    ] })
  ] })
] }) }) }), S = ({
  embedId: a,
  isAgentTyping: n,
  activeConversationId: s,
  isAgentMessageLoading: x,
  setIsAgentMessageLoading: v
}) => {
  const h = W(z), { context: k } = A(), { embedDefaultAgent: f, defaultAgent: o } = D(
    (t) => t.conversation
  ), { mutate: u } = q(
    R()
  ), [b, p] = j(!1), r = M(null), C = async (t) => {
    await u({
      message: {
        content: t
      },
      conversationId: s,
      agent: f[a] || o,
      context: k
    }), p(!0);
  }, [l, w] = j(""), i = (t) => {
    w(t.target.value);
  }, g = async (t) => {
    x || (await C(l), w(""), v(!0), t.preventDefault());
  }, H = (t) => {
    t.key === "Enter" && (n || !l) ? t.preventDefault() : t.key === "Enter" && !t.shiftKey && (t.preventDefault(), g(t), E !== null && (E.style.height = "auto"));
  }, N = async (t) => {
    if (!t || n)
      return null;
    x || (console.log("hit"), await C(t), w(""), v(!0)), E !== null && (E.style.height = "auto");
  }, E = typeof document < "u" ? document.querySelector("textarea") : null;
  return E && E.addEventListener("input", function() {
    this.style.height = "auto", this.style.height = this.scrollHeight + "px";
  }), m(() => {
    var t;
    b && ((typeof window < "u" ? window.innerWidth : 650) < 640 && ((t = r == null ? void 0 : r.current) == null || t.blur()), p(!1));
  }, [b]), /* @__PURE__ */ e.jsxs("div", { className: "border-t-1", children: [
    /* @__PURE__ */ e.jsxs("div", { className: "w-full min-h-14 bg-off-white flex items-center max-h-36 scrollbar-hidden py-4", children: [
      /* @__PURE__ */ e.jsx(
        "textarea",
        {
          placeholder: "Send a message...",
          ref: r,
          value: l,
          onChange: i,
          rows: 1,
          className: "font-sans h-auto px-4 max-h-36 w-full bg-off-white text-slate-500 scrollbar-hidden resize-none focus:outline-none focus:ring-blue-400 focus:ring-0",
          onKeyDown: H
        }
      ),
      n ? /* @__PURE__ */ e.jsx(G, { className: "mr-2 w-10 h-10 -mb-6 items-center" }) : /* @__PURE__ */ e.jsx(
        "button",
        {
          onClick: () => N(l),
          className: "bg-transparent",
          children: (h == null ? void 0 : h.sendMessageIcon) || /* @__PURE__ */ e.jsx(Q, { className: "h-6 w-6 mr-2", color: "black" })
        }
      )
    ] }),
    /* @__PURE__ */ e.jsxs("p", { className: "font-sans text-center text-xs text-gray-400 bg-neutral-200", children: [
      "Powered by",
      " ",
      /* @__PURE__ */ e.jsx(
        "a",
        {
          href: "https://convostack.ai",
          target: "_blank",
          className: "font-semibold",
          children: "ConvoStack.ai"
        }
      )
    ] })
  ] });
}, ee = ({
  embedId: a,
  style: n,
  customStyling: s
}) => {
  const x = W(z), { openConversation: v, context: h } = A(), { embedDefaultAgent: k, defaultAgent: f } = D(
    (r) => r.conversation
  ), { data: o, isFetching: u, isLoading: b } = K(
    R(),
    void 0,
    {
      staleTime: 0
    }
  ), p = o !== void 0 && (o == null ? void 0 : o.getConversations) !== null && (o == null ? void 0 : o.getConversations) !== void 0 && (o == null ? void 0 : o.getConversations.length) > 0 ? o.getConversations : [];
  return m(() => {
    !u && p.length === 0 && v(
      null,
      k[a] || f,
      h,
      a
    );
  }, [u]), /* @__PURE__ */ e.jsxs(e.Fragment, { children: [
    /* @__PURE__ */ e.jsxs(
      "div",
      {
        className: `h-14 py-4 ${s != null && s.headerColor ? "" : "bg-blue-gradient"} flex justify-between items-center`,
        style: { backgroundColor: s == null ? void 0 : s.headerColor },
        children: [
          /* @__PURE__ */ e.jsx(
            "div",
            {
              className: "left-0 hover:cursor-pointer",
              onClick: () => v(
                null,
                k[a] || f,
                h,
                a
              ),
              children: (x == null ? void 0 : x.createNewConversationIcon) || /* @__PURE__ */ e.jsx(
                J,
                {
                  className: "w-6 h-6 ml-3",
                  color: (s == null ? void 0 : s.iconsColor) || "white"
                }
              )
            }
          ),
          /* @__PURE__ */ e.jsx("div", { className: "flex mx-auto", children: /* @__PURE__ */ e.jsx(
            "p",
            {
              className: "font-sans font-semibold mx-auto",
              style: { color: (s == null ? void 0 : s.headerTextColor) || "white" },
              children: (s == null ? void 0 : s.headerText) || "ConvoStack Chat"
            }
          ) })
        ]
      }
    ),
    /* @__PURE__ */ e.jsx(
      "div",
      {
        className: "bg-white overflow-y-auto flex flex-col pb-4",
        style: n,
        children: b ? /* @__PURE__ */ e.jsx($, { className: "mx-auto mt-8" }) : /* @__PURE__ */ e.jsx(e.Fragment, { children: p.length !== 0 && p.map((r, C) => {
          var l, w;
          return /* @__PURE__ */ e.jsx(
            O,
            {
              title: r.title || "",
              headline: ((l = r.lastMessage) == null ? void 0 : l.content) || "",
              updatedAt: ((w = r.lastMessage) == null ? void 0 : w.createdAt) || "",
              conversationId: r.id,
              avatarUrl: r.agent.avatarUrl || "",
              embedId: a
            },
            C
          );
        }) })
      }
    )
  ] });
}, se = ({ embedId: a, customStyling: n }) => {
  const s = W(z), x = U(), { dropSubscription: v } = A();
  return /* @__PURE__ */ e.jsx("div", { className: "w-full", children: /* @__PURE__ */ e.jsxs(
    "div",
    {
      className: `w-full h-14 ${n != null && n.headerColor ? "" : "bg-blue-gradient"} flex flex-wrap items-center py-4`,
      style: { backgroundColor: n == null ? void 0 : n.headerColor },
      children: [
        /* @__PURE__ */ e.jsx(
          "div",
          {
            className: "left-0 hover:cursor-pointer",
            onClick: () => {
              x(
                V({
                  embedId: a,
                  value: !0
                })
              ), v(a);
            },
            children: (s == null ? void 0 : s.backArrowIcon) || /* @__PURE__ */ e.jsx(
              X,
              {
                className: "w-6 h-6 ml-3",
                color: (n == null ? void 0 : n.iconsColor) || "white"
              }
            )
          }
        ),
        /* @__PURE__ */ e.jsx("div", { className: "flex mx-auto", children: /* @__PURE__ */ e.jsx(
          "p",
          {
            className: "font-sans font-semibold mx-auto",
            style: { color: (n == null ? void 0 : n.headerTextColor) || "white" },
            children: (n == null ? void 0 : n.headerText) || "ConvoStack Chat"
          }
        ) })
      ]
    }
  ) });
}, ae = ({
  isAgentTyping: a,
  setIsAgentTyping: n,
  data: s,
  style: x,
  CustomMessage: v,
  isAgentMessageLoading: h,
  setIsAgentMessageLoading: k
}) => {
  const f = v || Z, [o, u] = j(!0), [b, p] = j("130px"), [r, C] = j(
    []
  ), [l, w] = j([]), [i, g] = j(!1), H = (d) => {
    var T;
    const c = (T = d.data) == null ? void 0 : T.subscribeConversationEvents;
    c.kind === "message_part" ? w((L) => [
      ...L,
      c.payload.chunk
    ]) : c.kind === "message" ? (c.payload.role === "AI" && k(!1), C((L) => [
      ...L,
      {
        content: c.payload.content,
        role: c.payload.role
      }
    ]), w([])) : c.kind === "conversation_metadata" && (C((L) => [
      ...L,
      { content: c.payload.primer, role: "AI" }
    ]), u(!1));
  };
  m(() => {
    s !== null ? (a && I("auto"), H(s)) : (u(!0), C([]));
  }, [s]);
  const N = M(), E = M(), [t, F] = j(0), I = (d) => {
    if (N.current.scrollTop < t)
      return;
    const c = N.current.clientHeight, T = E.current.clientHeight;
    N.current.scrollTo({
      top: T - c,
      left: 0,
      behavior: d
    }), t < N.current.scrollTop && F(N.current.scrollTop);
  };
  return m(() => {
    I("smooth");
  }, [r, i, l]), m(() => {
    l.length === 0 && a ? n(!1) : l.length > 0 && !a && (n(!0), k(!1), g(!1));
  }, [l.length]), m(() => {
    const d = () => {
      if (N.current) {
        const c = N.current.getBoundingClientRect().width.toString() + "px";
        b !== c && p(c);
      }
    };
    return d(), typeof window < "u" && window.addEventListener("resize", d), () => {
      typeof window < "u" && window.removeEventListener("resize", d);
    };
  }, []), m(() => {
    let d;
    return h && (d = setTimeout(() => {
      g(!0);
    }, 1500)), () => {
      clearTimeout(d);
    };
  }, [h]), /* @__PURE__ */ e.jsx("div", { ref: N, className: "bg-white h-full overflow-auto", style: x, children: /* @__PURE__ */ e.jsx("div", { ref: E, className: "flex flex-col", children: o ? /* @__PURE__ */ e.jsx($, { className: "mx-auto mt-12 z-0" }) : /* @__PURE__ */ e.jsxs(e.Fragment, { children: [
    r.map(
      (d, c) => d.content && /* @__PURE__ */ e.jsx(
        f,
        {
          width: b,
          message: { text: d.content, author: d.role },
          className: c === r.length - 1 && l.length === 0 ? "mb-3" : ""
        },
        c
      )
    ),
    h && i && /* @__PURE__ */ e.jsx(Y, { className: "ml-4 w-10 h-10 items-center" }),
    l.length !== 0 && /* @__PURE__ */ e.jsx(
      f,
      {
        width: b,
        message: { text: l.join(""), author: "AI" },
        className: "mb-3"
      }
    )
  ] }) }) });
}, ie = ({
  embedId: a,
  defaultAgent: n,
  customStyling: s,
  CustomMessage: x
}) => {
  const { graphqlUrl: v } = D(
    (i) => i.conversation
  ), [h, k] = j(!1), f = U(), o = M(), [u, b] = j(null), [p, r] = j(!1);
  m(() => {
    f(P({ embedId: a, value: null })), f(
      V({ embedId: a, value: !0 })
    ), f(_({ embedId: a, value: null })), f(B({ embedId: a, value: n }));
  }, [a]), m(() => {
    f(B({ embedId: a, value: n }));
  }, [a, n]);
  const C = D(
    (i) => i.conversation.embedActiveConversationId[a]
  ), l = D(
    (i) => i.conversation.isEmbedConversationListVisible[a]
  ), w = D(
    (i) => i.conversation.embedData[a]
  );
  return m(() => {
    const i = () => {
      if (o.current) {
        const g = o.current.getBoundingClientRect().height.toString() + "px";
        u !== g && b(g);
      }
    };
    return i(), typeof window < "u" && window.addEventListener("resize", i), () => {
      typeof window < "u" && window.removeEventListener("resize", i);
    };
  }, []), /* @__PURE__ */ e.jsx(
    "div",
    {
      ref: o,
      className: "convostack overflow-hidden",
      style: {
        width: (s == null ? void 0 : s.embedWidth) || "auto",
        height: (s == null ? void 0 : s.embedHeight) || "400px",
        flex: (s == null ? void 0 : s.embedFlex) || ""
      },
      children: v === "" ? /* @__PURE__ */ e.jsx(y, {}) : l ? /* @__PURE__ */ e.jsx(
        ee,
        {
          embedId: a,
          style: { height: `calc(${u} - 56px` },
          customStyling: s
        }
      ) : /* @__PURE__ */ e.jsxs(
        "div",
        {
          className: "flex flex-col",
          style: { height: (s == null ? void 0 : s.embedHeight) || "400px" },
          children: [
            /* @__PURE__ */ e.jsx(se, { embedId: a, customStyling: s }),
            /* @__PURE__ */ e.jsx(
              ae,
              {
                style: { height: `calc(${u} - 112px` },
                isAgentTyping: h,
                setIsAgentTyping: k,
                data: w,
                CustomMessage: x,
                isAgentMessageLoading: p,
                setIsAgentMessageLoading: r
              }
            ),
            /* @__PURE__ */ e.jsx(
              S,
              {
                embedId: a,
                isAgentTyping: h,
                activeConversationId: C,
                isAgentMessageLoading: p,
                setIsAgentMessageLoading: r
              }
            )
          ]
        }
      )
    }
  );
};
export {
  ie as default
};
